std/string.d

```diff
@@ -581,29 +581,90 @@ ptrdiff_t indexOf(Char)(const(Char)[] s, in dchar c, in size_t startIdx,
 }
 
 /++
-    Returns the index of the first occurrence of $(D sub) in $(D s). If $(D sub)
-    is not found, then $(D -1) is returned.
+    Searches for substring in $(D s).
 
-    $(D cs) indicates whether the comparisons are case sensitive.
+    Params:
+        s = string or ForwardRange of characters to search in correct UTF format
+        sub = substring to search for
+        cs = CaseSensitive.yes or CaseSensitive.no
+
+    Returns:
+        the index of the first occurrence of $(D sub) in $(D s). If $(D sub)
+        is not found, then $(D -1) is returned.
+        If the parameters are not valid UTF, the result will still
+        be in the range [-1 .. s.length], but will not be reliable otherwise.
   +/
-ptrdiff_t indexOf(Char1, Char2)(const(Char1)[] s, const(Char2)[] sub,
-        in CaseSensitive cs = CaseSensitive.yes) @trusted
-    if (isSomeChar!Char1 && isSomeChar!Char2)
+ptrdiff_t indexOf(Range, Char2)(Range s, const(Char2)[] sub,
+        in CaseSensitive cs = CaseSensitive.yes)
+    if (isForwardRange!Range && isSomeChar!(ElementEncodingType!Range) && isSomeChar!Char2)
 {
     import std.uni : toLower;
-    import std.algorithm : find;
-    const(Char1)[] balance;
-    if (cs == CaseSensitive.yes)
+    alias Char1 = Unqual!(ElementEncodingType!Range);
+
+    static if (isSomeString!Range)
     {
-        balance = std.algorithm.find(s, sub);
+        import std.algorithm : find;
+
+        const(Char1)[] balance;
+        if (cs == CaseSensitive.yes)
+        {
+            balance = std.algorithm.find(s, sub);
+        }
+        else
+        {
+            balance = std.algorithm.find!
+                ((a, b) => std.uni.toLower(a) == std.uni.toLower(b))
```

Comment:
Although tangential to this PR as it's from the old code, this isn't actually correct, as not all code points have 1:1 upper:lower mapping.